### DISCUSSION TITLE
[FE102]捕獲與冒泡for迴圈練習疑問(事件代理之前)

### DISCUSSION THREAD

**Discussion by IamTomIamQQ** [Lidemy/forum#257] [前端相關問題]
### 狀況
&emsp;&emsp;由於總覺得自己早先跟課自己練習很不紮實(對照上BE101與爬文後評估自己的結論)，因此決定一邊整理筆順便就練習盲打(先盲打再整理，然後驗證自己理解有無問題)。發現了幾個測試老半天無法處裡的疑問。雖然沒造成我卡關但總覺得放著問題查不出來好悶(又怕放著放著就忘了)  
### 疑問一
&emsp;&emsp;練習時宣告變數我大多已經養成用let的習慣了，因此起先打完了時候竟然沒出現作用域先跑完數字永遠都卡在3的狀況，讓我有點傻眼。( 想說怎麼會這樣子~~(OㄢO) ) 然後當我改成var之後跳出的數字是4更是讓我傻眼了。疑問大致如下:  

**Q1：** 為何得到的`i`會變成 3 呢? 按理說不應該是得到 2 嗎?  
**Q2：** 請問我寫的推論(插在內容中)是正確的嗎? 
>>(總覺得對作用域距離理解還很遠)   

以下是我寫的內容：  
```
<body>
  <!-- 多個按鈕 與事件監聽 不使用代理監聽的土炮-->
  <div class="out">
    <div class="inn">
      <a href="/.test">連結</a>
        <button class="btn" data-x="1">1</button>
        <button class="btn" data-x="2">2</button>
        <button class="btn" data-x="3">3</button>
    </div>
  </div> 
  <script>
    //debugger;
    const x = document.querySelectorAll('.btn')
    for(var i=0; i<x.length; i++){

                  //Q2: for裡面如果用let 這問題就迎刃而解?，但我們要故意用var就會變成如下這樣子

      console.log(i)                        //Ｑ1-1：依序得到 0 1 2 三行
      x[i].addEventListener('click',
      function(e){
                                                   //console.log(i+1)  Ｑ1-2:會得到4
        alert(i+1)
      })
    }
  </script>
</body>

```

**Comment by aszx87410** [comment]
這個我記得在講事件的時候似乎有提過，或者在 [JS201] JavaScript 進階：那些你一直搞不懂的地方似乎也有

總之呢，當你用 var 寫的時候，從頭到尾就只有一個 i 變數，而迴圈跑完以後 i 會變成 3，所以你點下按鈕的時候會出現 3+1=4
用 let 的話每一次迴圈都會產生一個新的 i 變數，所以不會有這問題

**Reply by IamTomIamQQ** [reply]
&emsp;&emsp; var 的部分我後來仔細想想終於大略搞懂了，當初應該是整個沒仔細聽，筆記給漏了。

&emsp;&emsp;let的部分 我還是沒搞懂，但沒關係我還會抽空爬文跟思索；因為我中間又故意用let來繼續土炮不做代理；結果搞出了個很怪異的東西弄得我自己覺得頭好大，但我想先整理整理跟爬文，反正不影響我自己當下進度。
